### PR TITLE
refactor(registerObserver): preserve `observer` reference

### PR DESCRIPTION
**Pull request by adnasa**
1. returns `observer` when calling `registerObserver`
2. supercedes #26 
3. extends tests

---

#### Tech debt

1. Missing possibility to mock `logToConsole`
2. consider using `takeRecords` or another option over `callback`
One problem I found with the current implementation of passing `callback` into `window.PerformanceObserver` is that we "leaking" or "imposing" external functions into the instantiation of `window.PerformanceObserver`, which makes it hard or rather unintuitive to mock and test. One option to consider over this is to return a promise resolving the entries of the observer, but this is anti-pattern to Observers in general. A more optimal solution is a follow up and continuation to returning the instance is having `callback` removed and allow consumers to `registerOberserver` to use [`takeRecords`](https://developer.mozilla.org/en-US/docs/Web/API/PerformanceObserver/takeRecords) which fulfills two purpose:
   - empty the store
   - process data. Given this; we give raw records back and let consumers for themselves decide how they want to process this (optimal but breaking change)

### CODE REVIEW

File: src/npm/hook.js [hunk 1]
```diff
@@ -29,38 +29,31 @@ import { generateDataFromMeasures } from '../shared/generate'
 
   NOTE: This should only be used in development mode.
 */
-const registerObserver = (params, callback) => {
-  params = params || {}
-
-  // TODO: Is there any way to polyfill this API ?
```
Comment (adnasa) [inline src/npm/hook.js]: this is a devtool for chrome + firefox. no need to polyfill.
Writing add-ons/extensions is a whole different problem

File: src/npm/hook.js [hunk 2]
```diff
@@ -29,38 +29,31 @@ import { generateDataFromMeasures } from '../shared/generate'
...
-      }
-
+      if (typeof callback === 'function') callback(measures)
       window.__REACT_PERF_DEVTOOL_GLOBAL_STORE__ = {
         measures,
         length: list.getEntries().length,
         rawMeasures: list.getEntries()
       }
-
-      // For logging to console
```
[the remark this answers is not in the document]
Comment (nitin42) [reply src/npm/hook.js]: Oops! Yup, totally.

**Comment by nitin42** [comment]
I think if we go for `observer.takeRecords()` then users of this hook will receive the raw measures (not parsed and aggregated), and hence this goes against the usage of this hook. The main purpose of this hook is to provide users with options to get parsed measures so that they can send it over to their analytics server. If we undo this, then we may wanna expose utility functions too along with `registerObserver` to parse the measures. And I'm afraid that this will increase the API surface area. I just wanted to keep things minimal. Let me know what you think!

**Comment by adnasa** [comment]
> The main purpose of this hook is to provide users with options to get parsed measures so that they can send it over to their analytics server.

This is exactly what I had in mind when I discussed `takeRecords`. There is nothing stopping us from returning the following shape upon instantiation:

```js
return {
  // raw oberserver with it's API
  observer,
  // list of entries that were already processed with
  // `generateDataFromMeasures` and `getReactPerformanceData`
  // by react-perf-devtool
  compiledMeasurements,
}
```

which would serve the two main purposes with this change:

1. a reference of `observer` is received
2. a list of measurements are received

---

hmm.. I realised now that yes, you're right. You're right because the problem with my suggestion is that consumer can not receive new measurements as soon as they are updated.

let's leave this as is. please review :)

**Comment by nitin42** [comment]
I'll review your PR soon (by the end of this week). I am just busy doing interviews for full-time gig. Although the code changes looks good to me but I'll still pull the changes manually on my machine for behavioural testing. Thank you for all the contribution you've made to this project. I really appreciate your help 😄

**Comment by adnasa** [comment]
... 24 days later ... 😆 🤣

File: src/npm/hook.js (hunk 1 above)
Comment (flxwu) [reply src/npm/hook.js]: see #52, we'll polyfill for older browsers if we somewhen find a proper polyfill for the API :)

File: src/npm/hook.js [hunk 3]
```diff
@@ -29,38 +29,25 @@ import { generateDataFromMeasures } from '../shared/generate'
...
-        logToConsole(params, measures)
-      }
-    })
-
-    observer.observe({
-      entryTypes: ['measure']
+      if (params.shouldLog) logToConsole(params, measures)
     })
+    observer.observe({ entryTypes: ['measure'] })
+    return observer
```
Comment (flxwu) [inline src/npm/hook.js]: maybe you should then also adapt the README as in #26 to make sure the user declares the observer in the global scope

**Pull request merged**